src/compiler/types.ts

```diff
@@ -8685,29 +8685,74 @@ export interface TransformationResult<T extends Node> {
  * A function that is used to initialize and return a `Transformer` callback, which in turn
  * will be used to transform one or more nodes.
  */
-export type TransformerFactory<T extends Node> = (context: TransformationContext) => Transformer<T>;
+export type TransformerFactory<TIn extends Node, TOut extends Node | undefined = TIn> = (context: TransformationContext) => Transformer<TIn, TOut>;
 
 /**
  * A function that transforms a node.
  */
-export type Transformer<T extends Node> = (node: T) => T;
+export type Transformer<TIn extends Node, TOut extends Node | undefined = TIn> = (node: TIn) => TOut;

```

Comment:
A `Transformer` shouldn't allow you to return `undefined`, and should always produce a `TIn`. Are there places where this doesn't match?

Reply:
My impression was that it was always possible to return a completely different node type from the transform API, but maybe that impression was wrong. See also: https://github.com/microsoft/TypeScript/pull/49929#discussion_r923853952

But, the codebase does compile when I leave `T` there, so maybe I fixed this somewhere internally and this could be restored `(node: T) => T`.

Reply:
Yeah, I guess this isn't right. I couldn't find any reason why it could be undefined. I will restore it.